### PR TITLE
[dlib] update to 19.10

### PR DESCRIPTION
**Pull request by jasjuang**
Tested both `vcpkg install dlib and dlib[*]` and everything compiles as usual. Note that dlib now export dll starting from 19.10, but they didn't handle the dll macros in the headers hence we need to add `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE` in the cmake options.

### CODE REVIEW

**Comment by alexkaratarakis** [comment]
Thanks!

**Comment by patrikhuber** [comment]
Is it possible that this broke dlib...? I just updated from 19.9 to 19.10 and I'm now getting linker errors in my project that uses dlib:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2001	unresolved external symbol USER_ERROR__inconsistent_build_configuration__see_dlib_faq_1	app	app.obj	1	
Error	LNK2001	unresolved external symbol USER_ERROR__inconsistent_build_configuration__see_dlib_faq_1	app	myapp.lib(app.obj)	1	
Error	LNK2001	unresolved external symbol DLIB_VERSION_MISMATCH_CHECK__EXPECTED_VERSION_19_10_0	app	app.obj	1	
Error	LNK2001	unresolved external symbol DLIB_VERSION_MISMATCH_CHECK__EXPECTED_VERSION_19_10_0	app	myapp.lib(app.obj)	1	
```

I read #1038 and
```
#ifndef _NDEBUG
#define ENABLE_ASSERTS 1
#endif
```
didn't help.

**Comment by patrikhuber** [comment]
I reverted this PR, installed 19.9, and it works completely fine again. So this PR must've broken something indeed.

Btw. I only tried the `Debug` configuration when I encountered the error with 19.10 - I didn't try `Release` or `RelWithDebugInfo`. So it's possible this is related to that, and maybe @jasjuang only tested this in `Release` configuration.

**Comment by patrikhuber** [comment]
dlib also doesn't seem to work in RelWithDebugInfo mode... this is 19.9 because as mentioned I couldn't get 19.10 to work but presumably the same problem occurs in 19.10:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in my-app.obj	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(entropy_decoder_kernel_2.cpp.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MD_DynamicRelease' in my-app.obj	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(entropy_decoder_kernel_2.cpp.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in my-app.obj	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(base64_kernel_1.cpp.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MD_DynamicRelease' in my-app.obj	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(base64_kernel_1.cpp.obj)	1	
Error	LNK2001	unresolved external symbol USER_ERROR__missing_dlib_all_source_cpp_file__OR__inconsistent_use_of_DEBUG_or_ENABLE_ASSERTS_preprocessor_directives_	my-app	C:\tmp\build-vs2017\my-app.obj	1	
Error	LNK2001	unresolved external symbol USER_ERROR__missing_dlib_all_source_cpp_file__OR__inconsistent_use_of_DEBUG_or_ENABLE_ASSERTS_preprocessor_directives_	my-app	C:\tmp\build-vs2017\digitaldentistry.lib(my-lib.obj)	1	
Error	LNK2019	unresolved external symbol __imp__invalid_parameter referenced in function "void * __cdecl std::_Allocate_manually_vector_aligned<struct std::_Default_allocate_traits>(unsigned __int64)" (??$_Allocate_manually_vector_aligned@U_Default_allocate_traits@std@@@std@@YAPEAX_K@Z)	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(base64_kernel_1.cpp.obj)	1	
Error	LNK2019	unresolved external symbol __imp__CrtDbgReportW referenced in function "void * __cdecl std::_Allocate_manually_vector_aligned<struct std::_Default_allocate_traits>(unsigned __int64)" (??$_Allocate_manually_vector_aligned@U_Default_allocate_traits@std@@@std@@YAPEAX_K@Z)	my-app	C:\tmp\build-vs2017\dlib_debug_64bit_msvc1913.lib(base64_kernel_1.cpp.obj)	1	
Error	LNK1120	3 unresolved externals	my-app	C:\tmp\build-vs2017\RelWithDebInfo\my-app.exe	1	
```

**Comment by patrikhuber** [comment]
@ras0219 I am really close to throwing in the towel with vcpkg. I keep hitting issue after issue with these libraries in vcpkg. Port files don't seem to be fully tested and maintained, and things keep breaking after updates. I appreciate the huge effort a lot and vcpkg is awesome in principle, but these packages, particularly things like OpenCV and dlib, don't have the most straight-forward CMake files, and it's just too much of an expectation that community-maintenance of these things are going to work (particularly if they don't work properly from the start). vcpkg is supposed to make things easier and save its users time, and users shouldn't have to spend hours in trying to deal with broken packages week after week, or spend hours fixing portfiles.

I apologise for the sort-of rant, but this is really frustrating.

In this one particularly issue, using the `RelWithDebugInfo` configuration is really not too far-fetched.

**Comment by davisking** [comment]
@patrikhuber What do you expect me to do about it?  The dlib cmake scripts are easy to use, if you use cmake it's simple.  You could also follow the instructions here http://dlib.net/compile.html.  It's not my fault vcpkg is broken.

**Comment by davisking** [comment]
I should point out that the errors you guys are getting in 19.10 are detecting configuration errors that really are errors.  I added these new checks in 19.10 because a lot of users don't understand C++'s one definition rule and constantly violate it.  It's not a dlib rule, it's a core C++ rule.  The problem from my point of view is that package maintainers or inexperienced C++ developers break C++'s ODR and then complain to me that their programs don't work.  *but this issue has nothing to with dlib*.  

So I've gone out of my way to make it very difficult to create a program with ODR violations regarding any part of dlib's source code.  Just because you aren't getting these errors in 19.9 doesn't mean you aren't creating ODR violations in that case as well.  You almost certainly are and your programs therefore, according to the C++ standard, have undefined behavior. 

**TL;DR: Package maintainers need to understand the C++ ODR and not violate it.**

**Comment by patrikhuber** [comment]
> @patrikhuber What do you expect me to do about it? The dlib cmake scripts are easy to use, if you use cmake it's simple. You could also follow the instructions here http://dlib.net/compile.html. It's not my fault vcpkg is broken.

@davisking: Thank you for the reply!
I thought that there was a chance that you might be interested in getting dlib to work properly in vcpkg. Maybe not, and that's fine! (we all have priorities and time-constraints).
I know that I can compile dlib "outside" of vcpkg on my own, and it works just fine. Thanks for the great library! I am posting here because I would like to use it from within vcpkg, because it makes dependency management both for me and for users and clients much easier (that is, if it worked...). But I totally understand that this is not your responsibility.

**Comment by jasjuang** [comment]
@patrikhuber I am sorry this update breaks debug mode for you. @alexkaratarakis could you revert this PR for now? dlib seems to be a bit different than other libraries and I will need some time to investigate this, and @patrikhuber's client needs the debug mode working right now.

**Comment by jasjuang** [comment]
@davisking Would you consider support DLL macros in the upstream? Currently if I compile dlib with `-DBUILD_SHARED_LIBS=ON`, it only generates DLL but no lib files to link to. I have to force export the definitions by `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE` so I have something to link to but I suspect this functionality is not fully working and hence causing the error @patrikhuber is seeing.

**Comment by davisking** [comment]
@jasjuang You should probably not build shared libraries. Why don't you use static libraries?

In any case, if you want to submit a PR to the main dlib repo that helps that's fine with me so long as the change is not disruptive to other dlib uses.

Also, I'm doubtful that dlib is different from other libraries in this way.  The issue is more that I put code in 19.10 that doesn't do anything other than cause ODR violating build environments to generate the error you are experiencing. What makes you confident ODR isn't being violated here? Just because code is built and seems to work in other situations doesn't mean ODR isn't being violated.

**Comment by jasjuang** [comment]
@davisking Thanks for the quick reply. To be honest with you, I am not sure in what scenario would an ODR violation caused by vcpkg if it wasn't there in the upstream. vcpkg is not adding extra code to dlib. Could you educated me how vcpkg could cause an ODR violation?

Well, one reason not to use static library is to have smaller downstream binary, and in some scenarios not having to update our client's binary and just the dynamic library as long as it's binary compatible.

**Comment by ras0219-msft** [comment]
First, @patrikhuber thank you very much for bringing this issue to our attention -- I know it's frustrating when things don't work and I want you to know that we really do appreciate your persistence.

@davisking we are confident that ODR isn't violated because vcpkg is designed from the ground up with ABI/ODR concerns in mind -- we build a single copy of each library into dedicated, individual folders and use those for the builds (no mixing).

I've also reproduced and found a fix for this issue; the root cause is (as @jasjuang implied) attempting to consume dlib as DLLs. On Windows, linking _variables_ (not functions) across DLL boundaries requires them to be marked as `__declspec(dllimport)`. For this reason, CMake's `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` only works for functions. It appears that the only _global variable_ that's shared between dlib and consumer code is the ODR detection machinery, which is why those are the only symbols missing.

This can be trivially fixed by adding `__declspec(dllexport)/__declspec(dllimport)` macros around the ODR machinery, however if @davisking is not interested in supporting DLLs then it may be better to simply force static linkage.

@davisking 2 more reasons for shared libraries: license concerns (I know dlib is under the boost license, but inter-license-interactions always seem very complicated) as well as enabling other libraries _above_ dlib to be built as shared. For the exact ODR concerns you mentioned, as soon as one library in a dependency graph requires static linking, every library above can no longer safely provide shared linkage.

**Comment by davisking** [comment]
I'm in general fine with shared linking.  However, I get a lot of questions from confused and novice C++ users who don't understand the implications of shared linking so I discourage it.  But it should work fine. 

I also don't see why you couldn't statically link dlib into dependent libraries.  I do this all the time. What is the issue here with it?

**Comment by davisking** [comment]
Also, if someone wants to submit a PR that exports the relevant variables I'm cool with that :)

**Comment by ras0219-msft** [comment]
The issue is that if you use dlib yourself, and a dependency of yours uses dlib, and it's static-link-only, you'll end up with two copies in your process

**Comment by davisking** [comment]
@ras0219-msft Like 99% of the code in dlib is in templates.  Worrying about duplicating the 1% that isn't is pointless since you are going to duplicate the other 99% anyway.

**Comment by davisking** [comment]
Sure, appveyor already runs lots of dlib tests.  I don't see why it can't do this as well.

**Comment by ras0219-msft** [comment]
>99% of the code in dlib is in templates.
Alright, that's a good point. In that case, it doesn't seem worth bothering with shared libraries.

@patrikhuber I've pushed a commit that disables shared linking to master and confirmed that the bridge_ex[1] example builds in both Release and Debug.

Unfortunately, RelWithDebInfo defaults to Debug libraries when presented with Release and Debug, which under MSVC are ABI incompatible -- that's what all the `_ITERATOR_DEBUG_LEVEL` noise is about. MSVC produces debug information as a separate file (`.pdb`) from the main executable, so there's really no reason to not produce it in release mode (and we do this for all the vcpkg-built libraries). This can be done by adding the `/Zi` and `/DEBUG` switches:
```cmake
add_compile_options(/Zi)
set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /DEBUG")
```

[1] https://github.com/davisking/dlib/blob/master/examples/bridge_ex.cpp

**Comment by patrikhuber** [comment]
@ras0219-msft Cool, thank you very much! I'm testing it right now.

I'm not sure I understand the second part of the paragraph about debug information. What needs to be modified? dlib, or my executable? The two lines of CMake are also global statements, so I think they modify the global options, which is an anti-pattern nowadays. I'd rather use target-based commands, if possible. (for the first that would be `target_compile_options`, for the later one I'm not sure but I'll find out).

What about the following approach, which seems to work well with OpenCV?
`set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)`

**Comment by ras0219-msft** [comment]
Yeah, that second approach looks way better! Thanks for pointing it out!

I'll look into it and see if we can apply it globally across vcpkg.

For completeness, but I think your approach may be better: assuming you're using `RelWithDebInfo` to get debug information for your (otherwise Release) build, you can instead just add debug information to your normal Release build by adding `/Zi` which enables debug information in the compiler and `/DEBUG` which enables creates the PDB in the linker. We already do this in Vcpkg for all your dependencies (dlib in this case), so you don't need to worry about them :).
